Conformance2.0 rm (#890)
finos/FDC3: toolbox/fdc3-conformance/Metadata-Tests.md

# Metadata & Instance Test Cases ![2.0](https://img.shields.io/badge/FDC3-2.0-blue)

You will need to pre-populate the AppDirectory with the following items:

| App | Required Metadata                        |
|-----|------------------------------------------|
| A   | Generic AppD Record which contains at least the following fields:<br>- `name`<br>- `version`<br>- `title`<br>- `tooltip`<br>- `description`<br>- `icons` (`Array<Icon>`)<br>- `screenshots` (`Array<Image>`)<br>- `interop.intents.listensFor` (`aTestingIntent` with at least context type `testContextX`)  |

> ## Get App Metadata
> 
> | App | Step           | Details                                                                                           |
> |-----|----------------|---------------------------------------------------------------------------------------------------|
> | Test   | 1.getAppMetadata    | Retrieve metadata for the configured app A with <br/> `const metadata1 = fdc3.getAppMetadata({appId: "<A's appId>"})`  |
> | Test   | 2.Confirm    | Compare the `AppMetadata` object to the expected definition for the fields provided above during setup and ensure that the metadata matches. An `instanceId` should NOT be set  |
> 
> - `GetAppMetadata`: perform the above steps

Reviewer: ```suggestion
## Retrieve `AppMetadata`

| App | Step           | Details                                                                                           |
|-----|----------------|---------------------------------------------------------------------------------------------------|
| Test   | 1.getAppMetadata    | Retrieve metadata for the configured app A with <br/> `const metadata1 = await fdc3.getAppMetadata({appId: "<A's appId>"})`  |
| Test   | 2.Confirm    | Compare the `AppMetadata` object to the expected definition for the fields provided above during setup and ensure that the metadata matches. An `instanceId` should NOT be set  |

- `GetAppMetadata`: perform the above steps
```

> ## Instance Metadata
> 
> | App | Step           | Details                                                                                           |
> |-----|----------------|---------------------------------------------------------------------------------------------------|
> | Test   | 1.Open1    | Open the first instance of App A using <br/> `const appIdentifier1 = await fdc3.open({appId: "<A's appId>"})` <br/>and retrieve its `AppIdentifier` with instance details.  |
> | Test   | 2.Open2    |Open a second instanceo of App A using <br>`const appIdentifier2 = await fdc3.open({appId: "<A's appId>"})` <br/> and retrieve its `AppIdentifier` with instance details.  Confirm that the `instanceId` differs from the first instance.  |
> | Test   | 3.getAppMetadata1    | Retrieve metadata for the first instance of the app with<br/> `const metadata1 = fdc3.getAppMetadata(appIdentifier1)` |
> | Test   | 4.Confirm1 | Compare the `AppMetadata` object to the expected definition for the fields provided above during setup and ensure that the metadata matches.  An `instanceId` should be provided, confirm that it matches the one in `appIdentifier1`  |
> | Test   | 5.getAppMetadata2    | Retrieve metadata for the second instance of the app with <br/>`const metadata1 = fdc3.getAppMetadata(appIdentifier2)`  |
> | Test   | 6.Confirm2    | An `instanceId` should be provided, confirm that it matches the one in `appIdentifier2`  |
> 
> - `AppInstanceMetadata`: Perform the above steps

Reviewer: Improve wording slightly 

```suggestion
## Instance Metadata

| App | Step           | Details                                                                                           |
|-----|----------------|---------------------------------------------------------------------------------------------------|
| Test   | 1.Open1    | Open a first instance of App A using <br/> `const appIdentifier1 = await fdc3.open({appId: "<A's appId>"})` <br/>and confirm that its `AppIdentifier` contains an `instanceId`.  |
| Test   | 2.Open2    |Open a second instance of App A using <br>`const appIdentifier2 = await fdc3.open({appId: "<A's appId>"})` <br/>and confirm that its `AppIdentifier` contains an `instanceId` and that its value differs from that returned for the first instance. |
| Test   | 3.getAppMetadata1    | Retrieve metadata for the first instance of the app with<br/> `const metadata1 = fdc3.getAppMetadata(appIdentifier1)` |
| Test   | 4.Confirm1 | Compare the `AppMetadata` object to the expected definition for the fields provided above during setup and ensure that the metadata matches.  |
| Test   | 5.getAppMetadata2    | Retrieve metadata for the second instance of the app with <br/>`const metadata1 = fdc3.getAppMetadata(appIdentifier2)`  |
| Test   | 6.Confirm2    | An `instanceId` should be provided, confirm that it matches the one in `appIdentifier2`  |

- `AppInstanceMetadata`: Perform the above steps
```

> ## Finding Instances
> 
> | App | Step           | Details                                                                                           |
> |-----|----------------|---------------------------------------------------------------------------------------------------|
> | Test   | 1.Open1    | Open the first instance of App A using <br/> `const appIdentifier1 = await fdc3.open({appId: "<A's appId>"})` <br/>and retrieve its `AppIdentifier` with instance details.  |
> | Test   | 2.Open2    |Open a second instanceo of App A using <br>`const appIdentifier2 = await fdc3.open({appId: "<A's appId>"})` <br/> and retrieve its `AppIdentifier` with instance details.  Confirm that the `instanceId` differs from the first instance.  |
> | Test   | 3.FindInstances    | Retrieve details of open instances with <br/> `let instances = await fdc3.findInstances({appId: "<A's appId>"})` <br/> confirm that both `appIdentifier1` and `appIdentifier2` are both present in the array.  |
> | Test   | 4.RaiseIntent   | Use `appIdentifier1` to raise an intent and target that instance, with<br/> `const resolution = fdc3.raiseIntent("aTestingIntent", {"type": "testContextX"}, appIdentifier1)` |
> | Test   | 5.Confirm1 | Check that `resolution.source` matches `appIdentifier1` |
> | A | 6.ConfirmReceipt | Ensure that the instance of app A represented by `appIdentifier1` received the raised intent |
> 
> - `FindInstances`: Perform the above steps

Reviewer: Improve wording slightly

```suggestion
## Finding Instances

| App | Step           | Details                                                                                           |
|-----|----------------|---------------------------------------------------------------------------------------------------|
| Test   | 1.Open1    | Open the first instance of App A using <br/> `const appIdentifier1 = await fdc3.open({appId: "<A's appId>"})` <br/>and confirm that its `AppIdentifier` contains an `instanceId`.  |
| Test   | 2.Open2    |Open a second instance of App A using <br>`const appIdentifier2 = await fdc3.open({appId: "<A's appId>"})` <br/>and confirm that its `AppIdentifier` contains an `instanceId` and that its value differs from that returned for the first instance. |
| Test   | 3.FindInstances    | Retrieve details of open instances of app A with <br/> `let instances = await fdc3.findInstances({appId: "<A's appId>"})` <br/> confirm that both `appIdentifier1` and `appIdentifier2` are both present in the array.  |
| Test   | 4.RaiseIntent   | Use `appIdentifier1` to raise an intent and target that instance, with<br/> `const resolution = fdc3.raiseIntent("aTestingIntent", {"type": "testContextX"}, appIdentifier1)` |
| Test   | 5.Confirm1 | Check that `resolution.source` matches `appIdentifier1` |
| A | 6.ConfirmReceipt | Ensure that the instance of app A represented by `appIdentifier1` received the raised intent |

- `FindInstances`: Perform the above steps
```

> ## Getting Info For The Agent
> 
> | App | Step           | Details                                                                                           |
> |-----|----------------|---------------------------------------------------------------------------------------------------|
> | Test   | 1.getInfo    |Retrieve the `ImplementationMetadata` for the DesktopAgent with <br/> `fdc3.getInfo().then((implMetadata) => {  ... subsequent steps ...}`<br>**Note that the use of `then` is deliberate and intended to confirm that a promise returned (as this function switched from synchronous to asynchronous in 2.0)**|
> | Test   | 2.CheckVersion  | Check that the `fdc3Version` variable is present and at or greater than 2.0 (which you can do with the [`versionIsAtLeast` function from FDC3's Methods.ts](https://github.com/finos/FDC3/blob/add64f8302c6dcdc8437cf0e245101e927b69ec2/src/api/Methods.ts#L207):<br>`const isFDC3v2 = versionIsAtLeast(implMetadata, "2.0")`  |
> | Test   | 3.CheckProvider  | Check that the `provider` variable is present and not an empty string  |
> | Test   | 4.CheckFeatures  | Check that the `optionalFeatures`, `optionalFeatures.OriginatingAppMetadata` and `optionalFeatures.UserChannelMembershipAPIs` variables are all present and that the latter two provide boolean values  |

Reviewer: Test needed in 1.2 as well, hence restructure slightly. 

Reading this one its clear that we need to use `.then` in other tests where the signature changed to async (as await passes through sync calls and will conceal an issue there).

```suggestion
## Getting Info For The Desktop Agent

| App | Step           | Details                                                                                           |
|-----|----------------|---------------------------------------------------------------------------------------------------|
| Test   | 1.getInfo    |Retrieve the `ImplementationMetadata` for the DesktopAgent with: <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green)  `let implMetadata = fdc3.getInfo()`<br/>![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `fdc3.getInfo().then((implMetadata) => {  ... subsequent steps ...}`<br>**Note that the use of `then` is deliberate and intended to confirm that a promise returned (as this function switched from synchronous to asynchronous in 2.0)**|
| Test   | 2.CheckVersion  | Check that the `fdc3Version` variable is present and at (as appropriate) 1.2 or >= 2.0 (which you can do with the [`versionIsAtLeast` function from FDC3's Methods.ts](https://github.com/finos/FDC3/blob/add64f8302c6dcdc8437cf0e245101e927b69ec2/src/api/Methods.ts#L207):<br>`const isFDC3v2 = versionIsAtLeast(implMetadata, "2.0")`  |
| Test   | 3.CheckProvider  | Check that the `provider` variable is present and not an empty string  |
| Test   | 4.CheckFeatures  | ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) Check that the `optionalFeatures`, `optionalFeatures.OriginatingAppMetadata` and `optionalFeatures.UserChannelMembershipAPIs` variables are all present and that the latter two provide boolean values  |
```

Author: Metadata tests currently only apply to 2.0
There is a `BasicGI1` test which calls the 1.2 getInfo() method, but doesn't assert a specific FDC3 version.

Do we want `GetInfo1` to also be a 1.2 test?  Because it isn't at the moment

Reviewer: It should be really, the API function was required and not optional...

Author: ok, done

> - `GetInfo1`: Perform the above steps

Reviewer: Test applies to both versions (apart from step 4)
```suggestion
- `GetInfo1`: ![1.2](https://img.shields.io/badge/FDC3-1.2-green) ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) Perform the above steps
```

Author: See previous comment - while I doubt this is going to cause anyone any trouble, it does change the 1.2 test pack.

Reviewer: As above it probably should go in, was a small oversight

> | App | Step           | Details                                                                                           |
> |-----|----------------|---------------------------------------------------------------------------------------------------|
> | Test   | 1.Open1    | Start an instance of App A with <br/>`const appIdentifier1 = await fdc3.open({appId: "<A's appId>"})` <br /> retrieve its `AppIdentifier` with instance details. Confirm that the `AppIdentifier` contains both an `appId` and `instanceId` |
> | A | 2.getInfo     |   Retrieve the `ImplementationMetadata` for the DesktopAgent with: <br  />`fdc3.getInfo().then((implMetadata) => {  ... subsequent steps ...}` <br/> This should include `AppMetadata` for the retrieving app. |
> | A + Test | 3.Confirm  | Check that `implMetadata.appMetadata` contains an `appId` and `instanceId` matching that retrieved in the first step (will require transmission of the details from A to Test or vice-versa). Also compare the `AppMetadata` object to the expected definition for the fields provided above during setup and ensure that the metadata matches. |
> 
> - `GetInfo2`: Perform the above steps.

Reviewer: Note this test only applies to 2.0. 

```suggestion
| App | Step           | Details                                                                                           |
|-----|----------------|---------------------------------------------------------------------------------------------------|
| Test   | 1.Open1    | Start an instance of App A with <br/>`const appIdentifier1 = await fdc3.open({appId: "<A's appId>"})` <br /> retrieve its `AppIdentifier` with instance details. Confirm that the `AppIdentifier` contains both an `appId` and `instanceId` |
| A | 2.getInfo     |   Retrieve the `ImplementationMetadata` for the DesktopAgent with: <br  />`fdc3.getInfo().then((implMetadata) => {  ... subsequent steps ...}` <br/> This should include an `appMetadata` field with the `AppMetadata` for the retrieving app. |
| A + Test | 3.Confirm  | Check that `implMetadata.appMetadata` contains an `appId` and `instanceId` matching those retrieved in the first step (will require transmission of the details from A to Test or vice-versa). Also compare the `AppMetadata` object to the expected definition for the fields provided above during setup and ensure that the metadata matches. |

- `GetInfo2`: ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) Perform the above steps.
```

Author: At the top of the Metadata-Tests.md file there is a big "2.0" badge, indicating that the whole lot only applies to 2.0
However, if we make `GetInfo` apply to both versions I'll change this and add the badges to the individual tests in the file

Reviewer: 👍

Author: @Gaganahluwalla you'll need to make this change on your side too

Reviewer 2: @robmoffat - Does it mean a new Advanced FDC3.getInfo Test case is needed in 1.2?

Author: yes, that's right
